task/sp_measure/src/main.rs

```diff
@@ -0,0 +1,114 @@
+use ringbuf::*;
+use sha2::{Digest, Sha256};
+use userlib::*;
+
+const READ_SIZE: usize = 256;
+
+const TRANSACTION_SIZE: u32 = 1024;
+
+task_slot!(SP_CTRL, swd);
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    ErrCnt(usize),
+    Addr(u32),
+    Start(u64),
+    End(u64),
+    Data([u8; READ_SIZE]),
+    ShaGood,
+    ShaBad,
+    None,
+}
+
+ringbuf!(Trace, 16, Trace::None);
+
+#[export_name = "main"]
+fn main() -> ! {
+    let mut err_cnt = 0;
+    loop {
+        let mut sha = Sha256::new();
+        let sp_ctrl = SpCtrl::from(SP_CTRL.get_task_id());
+
+        match sp_ctrl.setup() {
+            Err(_) => panic!(),
+            _ => (),
+        }
+
+        let mut data: [u8; READ_SIZE] = [0; READ_SIZE];
+
+        let start = sys_get_timer().now;
+        ringbuf_entry!(Trace::Start(start));
+        for addr in (FLASH_START..FLASH_END).step_by(READ_SIZE) {
+            if addr % TRANSACTION_SIZE == 0 {
+                loop {
+                    match sp_ctrl
+                        .read_transaction_start(addr, addr + TRANSACTION_SIZE)
+                    {
+                        Err(_) => {
+                            err_cnt += 1;
+                            let _ = sp_ctrl.setup();
+                            continue;
+                        }
+                        _ => break,
+                    }
+                }
+            }
+
+            data.fill(0);
+            loop {
+                match sp_ctrl.read_transaction(&mut data) {
+                    Err(_) => {
+                        ringbuf_entry!(Trace::Addr(addr));
+                        ringbuf_entry!(Trace::Data(data));
+                        loop {
+                            match sp_ctrl.setup() {
+                                Err(_) => continue,
+                                Ok(_) => {
+                                    err_cnt += 1;
+                                    match sp_ctrl.read_transaction_start(
+                                        addr,
+                                        addr + TRANSACTION_SIZE,
+                                    ) {
+                                        Err(_) => continue,
+                                        Ok(_) => break,
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    Ok(_) => break,
+                }
+            }
+
+            sha.update(&data);
+        }
+
+        let sha_out = sha.finalize();
+
+        let end = sys_get_timer().now;
+        ringbuf_entry!(Trace::End(end));
+        ringbuf_entry!(Trace::ErrCnt(err_cnt));
+        if sha_out.as_slice() == EXPECTED.as_slice() {
+            ringbuf_entry!(Trace::ShaGood);
+        } else {
+            ringbuf_entry!(Trace::ShaBad);
+            panic!();
+        }
+
+        // Wait for a notification that will never come, politer than
+        // busy looping forever
```

Comment:
Neat!  Should we be doing this elsewhere?

Reply:
Maybe? I didn't find any other tasks that don't want to loop. I expect this will actually be folded into real measurements once we figure out how that works.